File: code/smv-debugger/src/main/java/org/fbme/smvDebugger/visualization/BacktraceService.java
package org.fbme.smvDebugger.visualization;

import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import jetbrains.mps.project.MPSProject;
import org.fbme.lib.iec61499.declarations.CompositeFBTypeDeclaration;
import java.util.List;

import org.fbme.lib.iec61499.fbnetwork.FunctionBlockDeclarationBase;
import org.fbme.lib.iec61499.fbnetwork.PortPath;
import org.fbme.lib.st.expressions.VariableReference;
import org.fbme.smvDebugger.model.SystemItemValue;
import org.fbme.smvDebugger.model.SystemItem;
import java.util.ArrayList;
import org.fbme.lib.iec61499.fbnetwork.FBNetwork;
import org.fbme.lib.iec61499.fbnetwork.FunctionBlockDeclaration;
import org.fbme.lib.iec61499.declarations.BasicFBTypeDeclaration;
import java.util.Objects;
import java.util.Optional;
import org.fbme.lib.iec61499.declarations.EventDeclaration;
import org.fbme.lib.iec61499.declarations.AlgorithmBody;
import org.fbme.lib.st.statements.AssignmentStatement;
import org.fbme.lib.st.expressions.Variable;
import org.fbme.lib.iec61499.ecc.StateTransition;
import org.fbme.lib.iec61499.ecc.ECTransitionCondition;

public class BacktraceService {
  private final Map<String, Set<String>> graph = new HashMap<>();
  private final Set<String> visited = new HashSet<>();
  private final MPSProject project;
  private final CompositeFBTypeDeclaration compositeFb;

  public BacktraceService(final MPSProject project, final CompositeFBTypeDeclaration compositeFb) {
    this.project = project;
    this.compositeFb = compositeFb;
  }

  public List<String> getRelatedItemSimpleNames(final SystemItemValue itemValue) {
    graph.clear();
    visited.clear();

    final SystemItem item = itemValue.getItem();
    switch (item.getType()) {
      case EVENT_PORT:
        backtraceEvent(item.getFbName(), item.getItemName());
      case DATA_PORT:
        backtraceData(item.getFbName(), item.getItemName());
      case ECC:
        backtraceEccState(item.getFbName(), item.getItemName());
    }

    final Set<String> relatedItemSimpleNames = new HashSet<>();
    relatedItemSimpleNames.addAll(graph.keySet());
    graph.values().forEach(relatedItemSimpleNames::addAll);

    return new ArrayList<>(relatedItemSimpleNames);
  }

  private void backtraceEvent(final String curFbName, final String event) {
    this.project.getModelAccess().runReadAction(() -> {
      final FBNetwork fbNethwork = compositeFb.getNetwork();
      final List<FunctionBlockDeclaration> fbs = fbNethwork.getFunctionBlocks();
      final BasicFBTypeDeclaration curFb = (BasicFBTypeDeclaration) fbs.stream().filter(it -> Objects.equals(it.getName(), curFbName)).findFirst().get().getTypeReference().getTarget();
      final Optional<EventDeclaration> inputEventOpt = curFb.getInputEvents().stream().filter(eventD -> Objects.equals(eventD.getName(), event)).findFirst();
      if (inputEventOpt.isPresent()) {
        final EventDeclaration inputEvent = inputEventOpt.get();
        fbNethwork.getEventConnections().stream().filter(con -> {
          final String target = con.getTargetReference().getTarget().getPortTarget().getName();
          return Objects.equals(target, event);
        }).forEach(con -> {
          final String fbName = con.getSourceReference().getTarget().getFunctionBlock().getName();
          final String eventName = con.getSourceReference().getTarget().getPortTarget().getName();
          final String fullName = fbName + "." + eventName;

          if (!visited.contains(fullName)) {
            visited.add(fullName);
            graph.putIfAbsent(fullName, new HashSet<String>());
            graph.get(fullName).add(curFbName + "." + event);
            backtraceEvent(fbName, eventName);
          }
        });
      }
    });
  }

  private void backtraceData(final String curFbName, final String var) {
    this.project.getModelAccess().runReadAction(() -> {
      final FBNetwork fbNethwork = compositeFb.getNetwork();
      final List<FunctionBlockDeclaration> fbs = fbNethwork.getFunctionBlocks();
      final BasicFBTypeDeclaration curFb = (BasicFBTypeDeclaration) fbs.stream().filter(it -> Objects.equals(it.getName(), curFbName)).findFirst().get().getTypeReference().getTarget();
      curFb.getAlgorithms().forEach(algorithm -> ((AlgorithmBody.ST) algorithm.getBody()).getStatements().forEach(statement -> {
        if (statement instanceof AssignmentStatement) {
          final AssignmentStatement assignment = ((AssignmentStatement) statement);
          final Variable curVar = assignment.getVariable();
          if (curVar instanceof VariableReference) {
            if (Objects.equals(((VariableReference) curVar).getReference().getTarget().getName(), var)) {
              curFb.getEcc().getStates().forEach(state -> {
                if (state.getActions().stream().anyMatch(action -> Objects.equals(action.getAlgorithm().getTarget(), algorithm))) {
                  final String eccName = state.getName();
                  final String fullName = curFbName + "." + eccName;
                  if (!(visited.contains(fullName))) {
                    visited.add(fullName);
                    graph.putIfAbsent(fullName, new HashSet<>());
                    graph.get(fullName).add(curFbName + "." + eccName);
                    backtraceEccState(curFbName, eccName);
                  }
                }
              });

            }
          }
        }
      }));
    });
  }

  private void backtraceEccState(final String curFbName, final String state) {
    this.project.getModelAccess().runReadAction(() -> {
      final FBNetwork fbNetwork = compositeFb.getNetwork();
      final List<FunctionBlockDeclaration> fbs = fbNetwork.getFunctionBlocks();
      final BasicFBTypeDeclaration curFb = (BasicFBTypeDeclaration) fbs.stream().filter(it -> Objects.equals(it.getName(), curFbName)).findFirst().get().getTypeReference().getTarget();
      final List<StateTransition> transitions = curFb.getEcc().getTransitions();
      transitions.stream().filter(transition -> Objects.equals(transition.getTargetReference().getTarget().getName(), state)).forEach(transition -> {
        final ECTransitionCondition condition = transition.getCondition();
        PortPath<EventDeclaration> target = condition.getEventReference().getTarget();
        if (target != null) {
          FunctionBlockDeclarationBase functionBlock = target.getFunctionBlock();
          final String eventName = target.getPortTarget().getName();
          final String fullName = functionBlock == null ? eventName : functionBlock.getName() + "." + eventName;
          if (!(visited.contains(fullName))) {
            visited.add(fullName);
            graph.putIfAbsent(fullName, new HashSet<>());
            graph.get(fullName).add(curFbName + "." + state);
            backtraceEvent(curFbName, fullName);
          }
        }
      });
    });
  }
}

File: code/smv-debugger/src/main/kotlin/org/fbme/smvDebugger/visualization/BacktraceService.kt
package org.fbme.smvDebugger.visualization

import jetbrains.mps.project.MPSProject
import org.fbme.lib.iec61499.declarations.AlgorithmBody
import org.fbme.lib.iec61499.declarations.BasicFBTypeDeclaration
import org.fbme.lib.iec61499.declarations.CompositeFBTypeDeclaration
import org.fbme.lib.iec61499.declarations.EventDeclaration
import org.fbme.lib.iec61499.ecc.StateAction
import org.fbme.lib.iec61499.ecc.StateDeclaration
import org.fbme.lib.iec61499.ecc.StateTransition
import org.fbme.lib.iec61499.fbnetwork.FBNetworkConnection
import org.fbme.lib.iec61499.fbnetwork.FunctionBlockDeclaration
import org.fbme.lib.st.expressions.VariableReference
import org.fbme.lib.st.statements.AssignmentStatement
import org.fbme.smvDebugger.model.SystemItemType
import org.fbme.smvDebugger.model.SystemItemValue
import java.util.function.Consumer

class BacktraceService(private val project: MPSProject, private val compositeFb: CompositeFBTypeDeclaration) {
    private val graph: MutableMap<String, MutableSet<String>> = HashMap()
    private val visited: MutableSet<String> = HashSet()
    fun getRelatedItemSimpleNames(itemValue: SystemItemValue?): List<String> {
        graph.clear()
        visited.clear()
        val item = itemValue!!.item
        when (item.type) {
            SystemItemType.EVENT_PORT -> {
                backtraceEvent(item.fbName, item.itemName)
                backtraceData(item.fbName, item.itemName)
                backtraceEccState(item.fbName, item.itemName)
            }

            SystemItemType.DATA_PORT -> {
                backtraceData(item.fbName, item.itemName)
                backtraceEccState(item.fbName, item.itemName)
            }

            SystemItemType.ECC -> backtraceEccState(item.fbName, item.itemName)
        }
        val relatedItemSimpleNames: MutableSet<String> = HashSet()
        relatedItemSimpleNames.addAll(graph.keys)
        graph.values.forEach {
            relatedItemSimpleNames.addAll(it)
        }
        return ArrayList(relatedItemSimpleNames)
    }

    private fun backtraceEvent(curFbName: String?, event: String?) {
        project.modelAccess.runReadAction {
            val fbNethwork = compositeFb.network
            val fbs: List<FunctionBlockDeclaration> = fbNethwork.functionBlocks
            val curFb = fbs.stream().filter { it: FunctionBlockDeclaration -> it.name == curFbName }.findFirst()
                .get().typeReference.getTarget() as BasicFBTypeDeclaration?
            val inputEventOpt =
                curFb!!.inputEvents.stream().filter { eventD: EventDeclaration -> eventD.name == event }.findFirst()
            if (inputEventOpt.isPresent) {
                val inputEvent = inputEventOpt.get()
                fbNethwork.eventConnections.stream().filter { con: FBNetworkConnection ->
                    val target = con.targetReference.getTarget()!!.portTarget.name
                    target == event
                }.forEach { con: FBNetworkConnection ->
                    val fbName = con.sourceReference.getTarget()!!.functionBlock!!.name
                    val eventName = con.sourceReference.getTarget()!!.portTarget.name
                    val fullName = "$fbName.$eventName"
                    if (!visited.contains(fullName)) {
                        visited.add(fullName)
                        graph.putIfAbsent(fullName, HashSet())
                        graph[fullName]!!.add("$curFbName.$event")
                        backtraceEvent(fbName, eventName)
                    }
                }
            }
        }
    }

    private fun backtraceData(curFbName: String?, `var`: String?) {
        project.modelAccess.runReadAction {
            val fbNethwork = compositeFb.network
            val fbs: List<FunctionBlockDeclaration> = fbNethwork.functionBlocks
            val curFb = fbs.first { it.name == curFbName }.typeReference.getTarget() as BasicFBTypeDeclaration
            curFb.algorithms.forEach { algorithm ->
                (algorithm.body as AlgorithmBody.ST).statements.forEach { statement ->
                    if (statement is AssignmentStatement) {
                        val curVar = statement.variable
                        if (curVar is VariableReference) {
                            if (curVar.reference.getTarget()!!.name == `var`) {
                                curFb.ecc.states.forEach(
                                    Consumer { state: StateDeclaration ->
                                        if (state.actions.stream()
                                            .anyMatch { action: StateAction -> action.algorithm.getTarget() == algorithm }
                                        ) {
                                            val eccName = state.name
                                            val fullName = "$curFbName.$eccName"
                                            if (!visited.contains(fullName)) {
                                                visited.add(fullName)
                                                graph.putIfAbsent(fullName, HashSet())
                                                graph[fullName]!!.add("$curFbName.$eccName")
                                                backtraceEccState(curFbName, eccName)
                                            }
                                        }
                                    }
                                )
                            }
                        }
                    }
                }
            }
        }
    }

    private fun backtraceEccState(curFbName: String?, state: String?) {
        project.modelAccess.runReadAction {
            val fbNetwork = compositeFb.network
            val fbs: List<FunctionBlockDeclaration> = fbNetwork.functionBlocks
            val curFb = fbs.first { it.name == curFbName }.typeReference.getTarget() as BasicFBTypeDeclaration
            val transitions: List<StateTransition> = curFb.ecc.transitions
            transitions.asSequence()
                .filter { transition -> transition.targetReference.getTarget()!!.name == state }
                .forEach { transition ->
                    val condition = transition.condition
                    val target = condition.eventReference.getTarget()
                    if (target != null) {
                        val functionBlock = target.functionBlock
                        val eventName = target.portTarget.name
                        val fullName = if (functionBlock == null) eventName else functionBlock.name + "." + eventName
                        if (!visited.contains(fullName)) {
                            visited.add(fullName)
                            graph.putIfAbsent(fullName, HashSet())
                            graph[fullName]!!.add("$curFbName.$state")
                            backtraceEvent(curFbName, fullName)
                        }
                    }
                }
        }
    }
}
